docs: Verify/review config descriptions

maria-robobug opened the pull request:
- Reviewed and modified descriptions.
- Added some minor formatting.
- Removed the "Key:" before every key name, seems redundant.

maria-robobug committed:
CLIO-989 - Verify/review config descriptions

Updated the doc style a bit, and reviewed descriptions.

kuznetsss committed:
feat: Expose ledger cache full and disabled to prometheus (#1957)

Fixes #1771

kuznetsss commented:
@PeterChen13579, could you please bring these changes into code without updating the md file?

maria-robobug commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-- **Constraints**: The value must be one of the following: `cassandra`
- -   **Description**: Type of database to use. We currently support Cassandra and Scylladb. We default to Scylladb.
-### Key: database.cassandra.contact_points
+- **Default value**: `cassandra`
```
@PeterChen13579 is the default value cassandra or scylladb? Got conflicting info here as the description says ScyllaDB is default.

kuznetsss replied:
It is `cassandra` as a database type. Cassandra and Scylla have the same API.

kuznetsss commented:
@maria-robobug, could you please add a link to this file somewhere in docs?

maria-robobug commented:
> @maria-robobug, could you please add a link to this file somewhere in docs?

@kuznetsss good point. I could reference it from the page below:
https://github.com/XRPLF/clio/blob/develop/docs/configure-clio.md

I'll update the PR shortly...

maria-robobug committed:
Mention config descriptions doc in configure-clio.md

PeterChen13579 commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
+
+This document provides a list of all available Clio configuration properties in detail.
+
+> [!INFO]
```
I'm using https://markdownlivepreview.com/ and it feel like here looks a bit weird 🤔 
<img width="258" alt="Screenshot 2025-03-12 at 5 37 19 PM" src="https://github.com/user-attachments/assets/e1dde5bd-c285-4061-bd99-82a64217d3fc" />

maria-robobug replied:
Hey ah my bad, that should say [!NOTE] even. Also it's a Github flavoured markdown admonition, so it's possible that _markdownlivepreview_ would not support it.

maria-robobug replied:
Fixed from INFO > NOTE. @PeterChen13579, if you preview the markdown in Github, you'll see how the admonition is meant to render.

https://github.com/XRPLF/clio/blob/987a5764e61a1af1c1c75a23f114f06b8cda2b37/docs/config-description.md

PeterChen13579 replied:
I see, Thanks for the clarification! 😄

PeterChen13579 reviewed (commented):
Thanks for the review! 😃 Just one question

maria-robobug committed:
Fix admonition type

github-actions[bot] committed:
style: clang-tidy auto fixes (#1963)

Fixes #1962.

Co-authored-by: godexsoft <385326+godexsoft@users.noreply.github.com>

kuznetsss committed:
chore: Upgrade libxrpl to 2.4.0 (#1961)

PeterChen13579 committed:
add code to match description

PeterChen13579 committed:
optimize

kuznetsss committed:
chore: Add git-cliff config (#1965)

First step for #1779.

kuznetsss commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
- -   **Description**: Type of database to use. We currently support Cassandra and Scylladb. We default to Scylladb.
-### Key: database.cassandra.contact_points
+- **Default value**: `cassandra`
+- **Constraints**: The value must be one of the following: `cassandra`, `scylladb`.
```
This is not correct. Only `cassandra` is allowed.

kuznetsss commented on src/util/newconfig/ConfigDescription.hpp:
```diff
@@ -124,134 +128,145 @@ struct ClioConfigDescription {
-           .value = "Indicates if the server should run with admin privileges. Only one of local_admin or "
-                    "admin_password can be set."},
+           .value =
+               "Indicates if the server should run with admin privileges. Note that this setting cannot be enabled "
```
```suggestion
               "Indicates if requests from `localhost` are allowed to call Clio admin-only APIs . Note that this setting cannot be enabled "
```

kuznetsss commented on src/util/newconfig/ConfigDescription.hpp:
```diff
@@ -124,134 +128,145 @@ struct ClioConfigDescription {
+        KV{.key = "cache.num_diffs",
+           .value = "The number of diffs to cache. For more information, see the [README.md](../src/etl/README.md) "
+                    "under etl."},
+        KV{.key = "cache.num_markers", .value = "The number of markers to cache."},
```
```suggestion
        KV{.key = "cache.num_markers", .value = "The number of markers to use at one time to traverse the ledger."},
```

kuznetsss commented on src/util/newconfig/ConfigDescription.hpp:
```diff
@@ -124,134 +128,145 @@ struct ClioConfigDescription {
+           .value = "The number of diffs to cache. For more information, see the [README.md](../src/etl/README.md) "
+                    "under etl."},
+        KV{.key = "cache.num_markers", .value = "The number of markers to cache."},
+        KV{.key = "cache.num_cursors_from_diff", .value = "The number of cursors that are different."},
```
```suggestion
        KV{.key = "cache.num_cursors_from_diff", .value = "The number of diffs to use to generate cursors."},
```

kuznetsss commented on src/util/newconfig/ConfigDescription.hpp:
```diff
@@ -124,134 +128,145 @@ struct ClioConfigDescription {
+                    "subscription-based tasks from `rippled`."},
+        KV{.key = "graceful_period", .value = "The number of milliseconds the server waits to shutdown gracefully."},
+        KV{.key = "cache.num_diffs",
+           .value = "The number of diffs to cache. For more information, see the [README.md](../src/etl/README.md) "
```
```suggestion
           .value = "The number of diffs to use to generate cursors. For more information, see the [README.md](../src/etl/README.md) "
```

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
+- **Constraints**: The value must be one of the following: `cassandra`, `scylladb`.
+- **Description**: Specifies the type of database to use for storing and retrieving data needed by the Clio server. Supported options are Cassandra and ScyllaDB. If you don't provide a value, the Clio server uses ScyllaDB by default.
+
+### database.cassandra.contact_points
```
If you set `database.type` to `scylladb`, all the keys below must also be `database.scylladb`. It would be good to reflect that somehow.

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
 - **Constraints**: None
- -   **Description**: If True, no ETL nodes will run with Clio.
-### Key: etl_sources.[].ip
+- **Description**: If set to `True`, no ETL nodes will run with Clio.
```
This sounds wrong. This settings merely allows Clio to start without any ETL sources (rippled nodes) being available at startup.

kuznetsss commented on src/util/newconfig/ConfigDescription.hpp:
```diff
@@ -124,134 +128,145 @@ struct ClioConfigDescription {
+                    "under etl."},
+        KV{.key = "cache.num_markers", .value = "The number of markers to cache."},
+        KV{.key = "cache.num_cursors_from_diff", .value = "The number of cursors that are different."},
+        KV{.key = "cache.num_cursors_from_account", .value = "The number of cursors from an account."},
```
```suggestion
        KV{.key = "cache.num_cursors_from_account", .value = "The number of cursors to fetch from account."},
```

kuznetsss commented on src/util/newconfig/ConfigDescription.hpp:
```diff
@@ -124,134 +128,145 @@ struct ClioConfigDescription {
+        KV{.key = "cache.num_markers", .value = "The number of markers to cache."},
+        KV{.key = "cache.num_cursors_from_diff", .value = "The number of cursors that are different."},
+        KV{.key = "cache.num_cursors_from_account", .value = "The number of cursors from an account."},
+        KV{.key = "cache.page_fetch_size", .value = "The page fetch size for cache operations."},
```
```suggestion
        KV{.key = "cache.page_fetch_size", .value = "The number of ledger objects to fetch concurrently per marker."},
```

kuznetsss commented on src/util/newconfig/ConfigDescription.hpp:
```diff
@@ -124,134 +128,145 @@ struct ClioConfigDescription {
+                    "in logging, a new log file starts."},
+        KV{.key = "log_tag_style", .value = "The style for log tags."},
+        KV{.key = "extractor_threads", .value = "The number of extractor threads."},
         KV{.key = "read_only", .value = "Indicates if the server should have read-only privileges."},
```
```suggestion
        KV{.key = "read_only", .value = "Indicates if the server is allowed to write data to the database."},
```

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: num_markers
+- **Default value**: `0`
+- **Constraints**: The value must be a positive double number.
+- **Description**: Specifies the timeout duration (in seconds) for RPC requests.
```
Does not sound correct. I think the intention of this is to specify how long we keep responding with a cached response for the same request (e.g. server_info). Needs confirmation though

PeterChen13579 replied:
Yea you're right. confirmed 👍

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: dos_guard.max_requests
+- **Default value**: `20`
+- **Constraints**: The minimum value is `0`. The maximum value is `4294967295`.
+- **Description**: The maximum number of concurrent connections allowed by DOS guard.
```
Concurrent connections for a specific IP address. Probably it's logical but could be better to specify to leave no place for any confusion.

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: workers
+- **Default value**: `1`
+- **Constraints**: The value must be a positive double number.
+- **Description**: Interval in seconds for DOS guard to sweep or clear its state.
```
Sweep and Clear is the same thing in this context. Maybe use `sweep (clear)`?

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
 - **Constraints**: None
- -   **Description**: Indicates if the server should run with admin privileges. Only one of local_admin or admin_password can be set.
-### Key: server.admin_password
+- **Description**: Indicates if the server should run with admin privileges. Note that this setting cannot be enabled together with [server.admin_password](#serveradmin_password), you must choose one or the other.
```
Sounds wrong also. This is a boolean flag that indicates that requests from `127.0.0.1` are automatically granted admin permission. There is no setting to make the server run with admin privileges.

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-- **Constraints**: The minimum value is `0`. The maximum value is `65535`
- -   **Description**: Optional parameter, used only if processing_strategy `parallel`. It limits the number of requests for a single client connection that are processed in parallel. If not specified, the limit is infinite.
-### Key: server.ws_max_sending_queue_size
+- **Constraints**: The minimum value is `0`. The maximum value is `65535`.
```
What happens if i specify `0`? @kuznetsss

PeterChen13579 replied:
Changed to 1 as well.

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: prometheus.enabled
+- **Default value**: `1500`
+- **Constraints**: The minimum value is `0`. The maximum value is `4294967295`.
+- **Description**: The maximum size of the websocket sending queue.
```
Would be nice to explain in few words what this queue actually is

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
- -   **Description**: Maximum size of the websocket sending queue.
-### Key: prometheus.enabled
+- **Default value**: `1500`
+- **Constraints**: The minimum value is `0`. The maximum value is `4294967295`.
```
What 0 means in practice?

PeterChen13579 replied:
Discussed with @kuznetsss , doesn't make sense for the value to be `0`. Changed to `1` instead

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: cache.num_diffs
+- **Default value**: `10`
+- **Constraints**: The value must be a positive double number.
+- **Description**: The number of milliseconds the server waits to shutdown gracefully.
```
Would be nice to specify that Clio is killed if we can't shutdown gracefully under the specified value

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: cache.num_markers
+- **Default value**: `32`
+- **Constraints**: The minimum value is `0`. The maximum value is `65535`.
+- **Description**: The number of diffs to cache. For more information, see the [README.md](../src/etl/README.md) under etl.
```
This sounds wrong a bit but i don't know the correct wording either. Maybe @cindyyan317 knows

PeterChen13579 replied:
fixed 👍

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: cache.num_cursors_from_diff
+- **Default value**: `48`
+- **Constraints**: The minimum value is `0`. The maximum value is `65535`.
+- **Description**: The number of markers to cache.
```
Not sure this is "to cache" either. @cindyyan317 do you have better explanation what it does?

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: cache.num_cursors_from_account
+- **Default value**: `0`
+- **Constraints**: The minimum value is `0`. The maximum value is `65535`.
+- **Description**: The number of cursors that are different.
```
Again, not sure this is true. @cindyyan317 please clarify 🙇

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
 - **Constraints**: None
- -   **Description**: Format string for log messages.
-### Key: log_to_console
+- **Description**: The format string for log messages.
```
I'd suggest to link the docs that describe the format: https://beta.boost.org/doc/libs/1_83_0/libs/log/doc/html/log/tutorial/formatters.html

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: extractor_threads
+- **Default value**: `none`
+- **Constraints**: The value must be one of the following: `int`, `uint`, `null`, `none`, `uuid`.
+- **Description**: The style for log tags.
```
Could be useful to add a note why different styles exist and how `uint`/`int` is faster than `uuid`.

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
 - **Constraints**: None
- -   **Description**: Indicates if the server should have read-only privileges.
-### Key: txn_threshold
+- **Description**: Indicates if the server should have read-only privileges.
```
This seems a bit off. The setting indicates whether this instance of Clio is *not* allowed to participate in the ETL process as a writer node (writing ledger data to the DB)

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: start_sequence
+- **Default value**: `0`
+- **Constraints**: The minimum value is `0`. The maximum value is `65535`.
+- **Description**: The transaction threshold value.
```
This is not very descriptive - i have no clue what it does after reading this. I'd like some explanation what this is

PeterChen13579 replied:
Deleted as it is not used in code;

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
- -   **Description**: Starting ledger index.
-### Key: finish_sequence
+- **Constraints**: The minimum value is `0`. The maximum value is `4294967295`.
+- **Description**: Starting ledger index.
```
If specified, this is the ledger index Clio will start populating the DB from

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
- -   **Description**: Ending ledger index.
-### Key: ssl_cert_file
+- **Constraints**: The minimum value is `0`. The maximum value is `4294967295`.
+- **Description**: The ending ledger index.
```
Similarly to above, this needs more detail. Like "Clio will stop writing ledger data to the DB once the specified ledger index is ingested"

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-### Key: api_version.max
+- **Default value**: `1`
+- **Constraints**: The minimum value is `1`. The maximum value is `3`.
+- **Description**: The minimum API version to use.
```
"Allowed", not "to use". And similarly for `max`

godexsoft commented on docs/config-description.md:
```diff
@@ -1,452 +1,600 @@
-
+- **Default value**: `100`
+- **Constraints**: The minimum value is `0`. The maximum value is `4294967295`.
+- **Description**: The number of cursors each coroutine will scan.
```
Coroutine is an implementation detail - i'd suggest using "job" or some other generic word describing a running task

PeterChen13579 committed:
fix: ripple_flag logic in account lines (#1969)

fixes #1968

PeterChen13579 committed:
fix a few descriptions

PeterChen13579 committed:
fix rest of description besides cursors

PeterChen13579 committed:
fix one more

PeterChen13579 committed:
fix cache description

godexsoft committed:
feat: ETLng extensions (#1967)

For #1599 #1597

github-actions[bot] committed:
style: clang-tidy auto fixes (#1972)

Fixes #1971. 
Co-authored-by: godexsoft <385326+godexsoft@users.noreply.github.com>

PeterChen13579 committed:
change marker/cursor description

godexsoft commented on docs/config-description.md:
```diff
@@ -12,8 +12,8 @@ This document provides a list of all available Clio configuration properties in
 - **Default value**: `cassandra`
-- **Constraints**: The value must be one of the following: `cassandra`, `scylladb`.
-- **Description**: Specifies the type of database to use for storing and retrieving data needed by the Clio server. Supported options are Cassandra and ScyllaDB. If you don't provide a value, the Clio server uses ScyllaDB by default.
+- **Constraints**: The value must be one of the following: `cassandra`.
```
Hm, this is not true (right?) 😨 But i also don't know how to communicate that all other keys need to be `database.scylladb.*` if you specify `scylladb` here.

PeterChen13579 replied:
I think @kuznetsss said only cassandra is allowed? 🤔 
Should we allow scylladb too?

godexsoft replied:
Let's see what @kuznetsss will say.. I have a feeling we messed up a bit if we don't actually support `database.scylladb.*` atm.

godexsoft commented on docs/config-description.md:
```diff
@@ -365,47 +365,47 @@ This document provides a list of all available Clio configuration properties in
 - **Default value**: `32`
 - **Constraints**: The minimum value is `0`. The maximum value is `65535`.
-- **Description**: The number of diffs to cache. For more information, see the [README.md](../src/etl/README.md) under etl.
+- **Description**: The number of cursors generated is the number of changed objects in the latest `cache.num_diffs` number of ledgers. Deleted objects does not count. Cursors are workers that load the ledger cache from the position of markers concurrently. For more information, please read [README.md](../src/etl/README.md) under etl.
```
Probably etl should be ETL. Also `Deleted objects do not count.` sounds more correct, wdyt? Or maybe 
```
The number of cursors generated is the number of changed (without counting deleted) objects in the latest `cache.num_diffs` number of ledgers.
```

PeterChen13579 replied:
this sounds better yea 😄

godexsoft commented on docs/config-description.md:
```diff
@@ -445,7 +445,7 @@ This document provides a list of all available Clio configuration properties in
 - **Default value**: `%TimeStamp% (%SourceLocation%) [%ThreadID%] %Channel%:%Severity% %Message%`
 - **Constraints**: None
-- **Description**: The format string for log messages.
+- **Description**: The format string for log messages. The format is described here: https://beta.boost.org/doc/libs/1_83_0/libs/log/doc/html/log/tutorial/formatters.html.
```
I may have accidentally copied a link that is for a "beta". I guess the correct link is https://www.boost.org/doc/libs/1_83_0/libs/log/doc/html/log/tutorial/formatters.html

godexsoft commented on docs/config-description.md:
```diff
@@ -493,47 +493,39 @@ This document provides a list of all available Clio configuration properties in
 - **Default value**: None
 - **Constraints**: The minimum value is `0`. The maximum value is `4294967295`.
-- **Description**: Starting ledger index.
+- **Description**: If specified, the starting ledger where `Clio` will starts writing to database from.
```
Typo: `starts` -> `start`. Maybe `If specified, the ledger index Clio will start writing to the database from.`. Also, Clio is spelled without ticks (unlike `rippled`).

godexsoft commented on docs/config-description.md:
```diff
@@ -493,47 +493,39 @@ This document provides a list of all available Clio configuration properties in
 - **Default value**: None
 - **Constraints**: The minimum value is `0`. The maximum value is `4294967295`.
-- **Description**: The ending ledger index.
+- **Description**: If specified, the final ledger where `Clio` will write to database.
```
Maybe `If specified, the final ledger that Clio will write to the database.`?

godexsoft commented on src/util/newconfig/ConfigConstraints.hpp:
```diff
@@ -78,7 +78,7 @@ static constexpr std::array<char const*, 3> kLOAD_CACHE_MODE = {
  * @brief specific values that are accepted for database type in config.
  */
-static constexpr std::array<char const*, 2> kDATABASE_TYPE = {"cassandra", "scylladb"};
+static constexpr std::array<char const*, 1> kDATABASE_TYPE = {"cassandra"};
```
This should not be changed.. i imagine someone out there uses `scylladb` as their DB setting and this will break their configuration.

godexsoft commented on src/util/newconfig/ConfigDescription.hpp:
```diff
@@ -136,8 +136,7 @@ struct ClioConfigDescription {
-                    "Supported options are Cassandra and ScyllaDB. If you don't provide a value, the Clio server uses "
-                    "ScyllaDB by default."},
+                    "Only `Cassandra` is currently allowed."},
         KV{.key = "database.cassandra.contact_points",
```
So here we are hardcoding `database.cassandra`. This may mean that we either need to copy the entire thing for `database.scylladb` or support it some other way by modifying the config framework.
@kuznetsss wdyt?

PeterChen13579 committed:
fix comments

cindyyan317 committed:
feat: Snapshot import feature (#1970)

Implement snapshot import cmd
`clio_snapshot --server --grpc_server 0.0.0.0:12345 --path
<snapshot_path>`

Implement snapshot range cmd
`./clio_snapshot --range --path <snapshot_path>`

Add
LedgerHouses: It is responsible for reading/writing snapshot data
Server: Start grpc server and ws server

kuznetsss commented on docs/config-description.md:
```diff
@@ -1,452 +1,592 @@
- -   **Description**: Maximum number of outstanding write requests. Write requests are api calls that write to database 
-### Key: database.cassandra.max_read_requests_outstanding
+- **Default value**: `10000`
+- **Constraints**: The minimum value is `0`. The maximum value is `4294967295`.
```
I think here and in many other places `0` as a minimum value doesn't make sense. Maybe we should switch the default minimum for number constaint to `1`? Do we have any place where `0` is a valid number?

PeterChen13579 committed:
fix min constraint value

kuznetsss committed:
ci: Pin cmake 3.31.6 for macos runners (#1983)

Fixes #1982.

kuznetsss committed:
fix: Fix ssl in new webserver (#1981)

Fixes #1980.

An SSL handshake was missing and WsConnection should be build from
stream not socket because in case SSL connection stream already
completed SSL handshake.

github-actions[bot] committed:
style: clang-tidy auto fixes (#1985)

Fixes #1984.

Co-authored-by: godexsoft <385326+godexsoft@users.noreply.github.com>

PeterChen13579 committed:
fix: incorrect set HighDeepFreeze flag (#1978)

fixes #1977

godexsoft committed:
chore: Update maintainers (#1987)

github-actions[bot] committed:
style: clang-tidy auto fixes (#1989)

Fixes #1988. Please review and commit clang-tidy fixes.

Co-authored-by: godexsoft <385326+godexsoft@users.noreply.github.com>

godexsoft commented on docs/config-description.md:
```diff
@@ -445,7 +445,7 @@ This document provides a list of all available Clio configuration properties in
 - **Default value**: `%TimeStamp% (%SourceLocation%) [%ThreadID%] %Channel%:%Severity% %Message%`
 - **Constraints**: None
-- **Description**: The format string for log messages. The format is described here: https://beta.boost.org/doc/libs/1_83_0/libs/log/doc/html/log/tutorial/formatters.html.
+- **Description**: The format string for log messages. The format is described here: https://www.boost.org/doc/libs/1_87_0/libs/log/doc/html/log/tutorial/formatters.html.
```
We don't use 1.87 yet (sadly). So i suggest we link to the version we actually use right now 👍

godexsoft reviewed (changes requested):
It's nearly there 👍

PeterChen13579 committed:
fix boost format link

PeterChen13579 committed:
change .md too

godexsoft committed:
chore: Update workflow notification settings (#1990)

godexsoft committed:
feat: ETLng integration (#1986)

For #1594

github-actions[bot] committed:
style: clang-tidy auto fixes (#1992)

Fixes #1991. 

Co-authored-by: godexsoft <385326+godexsoft@users.noreply.github.com>

kuznetsss committed:
feat: Nodes communication via DB (#1976)

Fixes #1966.

github-actions[bot] committed:
style: clang-tidy auto fixes (#1994)

Fixes #1993. 

Co-authored-by: godexsoft <385326+godexsoft@users.noreply.github.com>

godexsoft committed:
chore: Fix clang-tidy 'fixes' (#1996)

Undo some of #1994
